controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java

```diff
@@ -172,15 +178,19 @@ protected void startUp() {
                                                     .build();
 
             connectionFactory = new ConnectionFactoryImpl(clientConfig);
-            SegmentHelper segmentHelper = new SegmentHelper();
 
             AuthHelper authHelper = new AuthHelper(serviceConfig.getGRPCServerConfig().get().isAuthorizationEnabled(),
                     serviceConfig.getGRPCServerConfig().get().getTokenSigningKey());
-            streamMetadataTasks = new StreamMetadataTasks(streamStore, bucketStore, hostStore, taskMetadataStore,
-                    segmentHelper, controllerExecutor, host.getHostId(), connectionFactory, authHelper, requestTracker);
+            
+            segmentHelperRef.compareAndSet(null, new SegmentHelper(hostStore, connectionFactory, authHelper));
```

Comment:
@tkaitchuck 
> Why does segmentHelperRef need to exist? It is only used here.
This could just as easily read `SegmentHelper segmentHelper = new SegmentHelper(hostStore, connectionFactory, authHelper);`

It is used for testing. 
If test uses the constructor where it passes a segment helper instance (usually a mock to mock pravega tables), then that is used, otherwise a new instance is created during the initialization of the service. 

Note: we cannot create segment helper in the constructor because we ay not have all the other object available

Reply:
> Note: we cannot create segment helper in the constructor because we ay not have all the other object available

This is concerning because it implies a circular dependency. This shouldn't be possible if we are following our thread safety guidelines. Somewhere something is wrong. I don't know the code well enough to point out what.

Reply:
Its not circular dependency here. 
segmenthelper depends on hoststore connectionFactory and authHelper. 
ControllerService main creates them and then creates segmentHelper. 

However, for testing we want to use segmentHelperMock that does not depend on either of the three objects (basically doesnt talk to segment store) but has an inmemory table implementation.

So ControllerServiceMain tests we want to set segmentHelper with this mock instead of having the class create these objects and then a segmentHelper object that uses those objects.